Test functions 236

vishal-biyani opened the pull request:
- This is the first implementation of #236, tested with NodeJS so far. Ran into some issue with other env. will try again in a bit
- Changes from #259 have been cherry picked for this one
- Is still dependent on ```FISSION_ROUTER``` env as #242 is not merged yet

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fission/fission/355)
<!-- Reviewable:end -->

soamvasani commented on fission/function.go:
```diff
@@ -432,3 +433,69 @@ func fnPods(c *cli.Context) error {
+	fnReachable := false
+
+	for i := 0; i < maxRetries; i++ {
+		time.Sleep(retryDelay)
```
No need to sleep before the first request?

vishal-biyani replied:
Instead of a fixed sleep, I am retrying. Would a fixed sleep be good idea?

soamvasani replied:
The retry is good, but the first request attempt can be sent without waiting (in other words the sleep should be after the request attempt)

soamvasani commented on fission/function.go:
```diff
@@ -432,3 +433,69 @@ func fnPods(c *cli.Context) error {
+		time.Sleep(retryDelay)
+		resp = httpRequest(c.String("method"), url, c.String("body"), c.StringSlice("header"))
+		if resp.StatusCode < 300 {
+			body, _ := ioutil.ReadAll(resp.Body)
```
Don't ignore errors

soamvasani commented on fission/function.go:
```diff
@@ -432,3 +433,69 @@ func fnPods(c *cli.Context) error {
+
+	if !fnReachable {
+		fmt.Print("Function not reachable \n")
+		body, _ := ioutil.ReadAll(resp.Body)
```
handle errors

soamvasani commented on fission/function.go:
```diff
@@ -432,3 +433,69 @@ func fnPods(c *cli.Context) error {
+	if !fnReachable {
+		fmt.Print("Function not reachable \n")
+		body, _ := ioutil.ReadAll(resp.Body)
+		fmt.Print(string(body))
```
print status code too? Something like `Error calling function <function>: <code> <body>`

soamvasani commented on fission/function.go:
```diff
@@ -432,3 +433,69 @@ func fnPods(c *cli.Context) error {
+	//client := getClient(c.GlobalString("server"))
+	fnName := c.String("name")
+
+	fnCreate(c)
```
suggest removing

vishal-biyani replied:
I had thought that we wanted an end to end flow: create-test-delete but rechecked the description, will remove it.

soamvasani commented on fission/function.go:
```diff
@@ -432,3 +433,69 @@ func fnPods(c *cli.Context) error {
+		fnLogs(c)
+	}
+
+	fnDelete(c)
```
suggest removing

soamvasani commented on fission/main.go:
```diff
@@ -55,6 +58,8 @@ func main() {
 		{Name: "logs", Usage: "Display function logs", Flags: []cli.Flag{fnNameFlag, fnPodFlag, fnFollowFlag, fnDetailFlag, fnLogDBTypeFlag}, Action: fnLogs},
 		{Name: "pods", Usage: "Display function pods", Flags: []cli.Flag{fnNameFlag, fnLogDBTypeFlag}, Action: fnPods},
+		{Name: "invoke", Usage: "Invoke a function", Flags: []cli.Flag{fnNameFlag, fnMethodFlag, fnBodyFlag, fnHeaderFlag}, Action: fnInvoke},
+		{Name: "test", Usage: "Test a function", Flags: []cli.Flag{fnNameFlag, fnEnvNameFlag, fnCodeFlag, fnPackageFlag, fnSrcPackageFlag, htMethodFlag}, Action: fnTest},
```
I don't think we need both. Just test (or just invoke) is fine, with all the functionality in one command.

soamvasani commented on fission/function.go:
```diff
@@ -432,3 +433,69 @@ func fnPods(c *cli.Context) error {
+		body, _ := ioutil.ReadAll(resp.Body)
+		fmt.Print(string(body))
+		defer resp.Body.Close()
+		fnLogs(c)
```
don't ignore errors

vishal-biyani commented on fission/common.go:
```diff
@@ -50,3 +55,36 @@ func checkErr(err error, msg string) {
+		method != http.MethodDelete &&
+		method != http.MethodPost &&
+		method != http.MethodPut {
+		// Fatal causes some stale data for the calling function, needs improvement
```
Above is not valid now, my assumption of workflow was different. So removed relevant sections.

soamvasani commented:
@vishal-biyani Mind rebasing + addressing conflicts?

vishal-biyani commented:
@soamvasani Rebased + Solved merge conflicts.

soamvasani commented on fission/function.go:
```diff
@@ -134,6 +138,41 @@ func getContents(filePath string) []byte {
+	req, err := http.NewRequest("POST", queryURL.String(), nil)
+	checkErr(err, "Error creating logs request")
+
+	httpClient := http.Client{Timeout: 5 * time.Second}
```
How did you choose the timeout? Is it what kubectl does?

soamvasani commented on fission/function.go:
```diff
@@ -134,6 +138,41 @@ func getContents(filePath string) []byte {
+	if ldata == nil {
+		return false
+	}
+	fmt.Println("Data=", string(ldata))
```
Remove "Data="

soamvasani commented on fission/function.go:
```diff
@@ -134,6 +138,41 @@ func getContents(filePath string) []byte {
+	body, err := ioutil.ReadAll(resp.Body)
+	var data interface{}
+	json.Unmarshal(body, &data)
+	ldata, err := base64.StdEncoding.DecodeString(data.(string))
```
I'm confused about this. What does the body look like? Why use json.Unmarshal for just a string? Can you simply do `base64.StdEncoding.Decode(body)`?

soamvasani commented on fission/function.go:
```diff
@@ -134,6 +138,41 @@ func getContents(filePath string) []byte {
+	json.Unmarshal(body, &data)
+	ldata, err := base64.StdEncoding.DecodeString(data.(string))
+
+	if ldata == nil {
```
when would this happen?

soamvasani commented on fission/function.go:
```diff
@@ -497,3 +536,43 @@ func fnPods(c *cli.Context) error {
+
+	url := fmt.Sprintf("http://%s/fission-function/%s", routerURL, fnName)
+
+	maxRetries := 3
```
I don't think this command needs retries, let's keep it simple and the user can retry if they want.

soamvasani commented on fission/function.go:
```diff
@@ -497,3 +536,43 @@ func fnPods(c *cli.Context) error {
+
+	if !fnReachable {
+		body, err := ioutil.ReadAll(resp.Body)
+		checkErr(err, "Function test failed")
```
See other uses of checkErr -- this doesn't fit the grammar of the error message. Also, this error is not about the function test failing, it's about failing to read the http response

soamvasani commented on fission/function.go:
```diff
@@ -497,3 +536,43 @@ func fnPods(c *cli.Context) error {
+		checkErr(err, "Function test failed")
+		fmt.Printf("Error calling function %v: %v %v", fnName, resp.StatusCode, string(body))
+		defer resp.Body.Close()
+		flag := getPodLogs(c)
```
Don't use `bool` for error checking in Go -- use an `error`.

soamvasani commented on fission/function.go:
```diff
@@ -134,6 +138,41 @@ func getContents(filePath string) []byte {
 	return code
 }
 
+func getPodLogs(c *cli.Context) bool {
```
This function doesn't just get logs, it prints them too -- it should be called "show pod logs" or "print pod logs" or something like that

soamvasani commented on controller/functionApi.go:
```diff
@@ -188,3 +193,62 @@ func (a *API) FunctionLogsApiPost(w http.ResponseWriter, r *http.Request) {
+
+	// Get Unmanaged Pods first
+	selector := "unmanaged=true,functionName=" + fnName
+	podList, err := clientset.Core().Pods(ns).List(metav1.ListOptions{LabelSelector: selector})
```
Always need to handle errors!

soamvasani commented on controller/functionApi.go:
```diff
@@ -188,3 +193,62 @@ func (a *API) FunctionLogsApiPost(w http.ResponseWriter, r *http.Request) {
+	})
+
+	podLogOpts := v1.PodLogOptions{Container: envName} // Only the env container, not fetcher
+	podLogsReq := clientset.Core().Pods(ns).GetLogs(pods[0].ObjectMeta.Name, &podLogOpts)
```
`pods[0]` won't exist if there are no pods, and you'll get a runtime range check panic. So you have to handle the case where there are zero pods.

soamvasani commented on fission/common.go:
```diff
@@ -50,3 +55,37 @@ func checkErr(err error, msg string) {
+	}
+
+	req, err := http.NewRequest(method, url, strings.NewReader(body))
+	checkErr(err, "Error creating reuquest to function")
```
checkErr(err, "create HTTP request")

checkErr prints "Failed to <msg>" where msg is the second argument.

soamvasani commented on fission/function.go:
```diff
@@ -134,6 +138,41 @@ func getContents(filePath string) []byte {
+	}
+
+	queryURL, err := url.Parse(c.GlobalString("server"))
+	checkErr(err, "Error parsing the base URL")
```
error message grammar, `checkErr(err, "parse server URL")`

soamvasani commented on fission/function.go:
```diff
@@ -134,6 +138,41 @@ func getContents(filePath string) []byte {
+	queryURL.Path = fmt.Sprintf("/proxy/logs/%s", fnName)
+
+	req, err := http.NewRequest("POST", queryURL.String(), nil)
+	checkErr(err, "Error creating logs request")
```
error message grammar

vishal-biyani commented:
Hi @soamvasani I have addressed all comments except mechanism to find the "latest pod" for logs. For that one, I tried a few things, but the current logic seems to work based on sort in place. LMK and we can discuss it on slack if needed.

soamvasani commented on fission/function.go:
```diff
@@ -497,3 +528,36 @@ func fnPods(c *cli.Context) error {
+
+	url := fmt.Sprintf("http://%s/fission-function/%s", routerURL, fnName)
+
+	fnReachable := false
```
this function could be written more simply without this flag, right?

soamvasani commented on fission/function.go:
```diff
@@ -497,3 +528,36 @@ func fnPods(c *cli.Context) error {
+	fnReachable := false
+
+	resp := httpRequest(c.String("method"), url, c.String("body"), c.StringSlice("header"))
+	if resp.StatusCode < 300 {
```
3xx are not really errors, right? I don't think we should call them errors. This should be 400...

soamvasani commented on fission/function.go:
```diff
@@ -497,3 +528,36 @@ func fnPods(c *cli.Context) error {
+		checkErr(err, "Function test")
+		fmt.Print(string(body))
+		defer resp.Body.Close()
+		fnReachable = true
```
just return here instead of setting a flag?

soamvasani commented on fission/function.go:
```diff
@@ -134,6 +137,34 @@ func getContents(filePath string) []byte {
+
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return errors.New("get logs from POD directly")
```
s/POD/pod/

soamvasani commented on controller/functionApi.go:
```diff
@@ -188,3 +194,66 @@ func (a *API) FunctionLogsApiPost(w http.ResponseWriter, r *http.Request) {
+	}
+	envName := f.Spec.Environment.Name
+
+	_, clientset, err := tpr.GetKubernetesClient()
```
It would be better to move this to the init of the api object -- so you don't have to login to kubernetes api on every logs request

soamvasani commented on fission/function.go:
```diff
@@ -497,3 +528,36 @@ func fnPods(c *cli.Context) error {
+		checkErr(err, "read log response from pod")
+		fmt.Printf("Error calling function %v: %v %v", fnName, resp.StatusCode, string(body))
+		defer resp.Body.Close()
+		err = printPodLogs(c)
```
This logic -- calling print pod logs and falling back to the DB -- would be useful in the `fission fn logs` command as well.  (But we can push that to a later change if you want to.)

vishal-biyani replied:
@soamvasani I can integrate this and make it so that ``` fn logs``` calls pod logs and if not found then looks up DB logs. But there are two things which are inconsistencies from user perspective and we should address in future: 1) The logs from POD don't have a timestamp before every line while ```fn logs``` prints TS 2) The ```fn logs``` shows lot many lines as opposed to pod logs which shows only recent pod's logs. So the behavior/experience can vary based on if you test a function and call logs or call it without testing it. There might be some optimizations in fetching logs from DB side too. What do you think?

vishal-biyani commented on controller/functionApi.go:
```diff
@@ -188,3 +194,66 @@ func (a *API) FunctionLogsApiPost(w http.ResponseWriter, r *http.Request) {
+	ns := vars["namespace"]
+
+	if len(ns) == 0 {
+		ns = "fission-function"
```
@soamvasani I will create an issue so we can track this.

vishal-biyani commented:
Thanks @soamvasani for extensive review on first PR :) I have one comment specific to using new logs feature in ```fn logs```, based on your thoughts I will create a issue or fix in existing PR itself.

vishal-biyani commented:
Review status: 0 of 6 files reviewed at latest revision, 39 unresolved discussions.

---

*[controller/functionApi.go, line 29 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r6-29:-Kxg5dDiegjmZEoBAOjd:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/controller/functionApi.go#L29)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

skip this blank link please, we're following a convention of three "paragraphs" of imports: [standard libs] [external deps] [internal deps]
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 33 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r6-33:-Kxg2Pf1DknkWXh6B_ej:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/controller/functionApi.go#L33)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

commented out, please remove
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 221 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r6-221:-Kxg38koO6nc00jXvOcV:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/controller/functionApi.go#L221)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

You can just use `functionName`, right? No need for matching on `unmanaged`, too? (In the future we will have function pods inside deployments, and so without the unmanaged label.)
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 222 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r6-222:-Kxg39QE6WwpkApJPEcG:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/controller/functionApi.go#L222)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

Always need to handle errors! 
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 224 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r6-224:-Kxg39ztFVcw9t1PjZZd:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/controller/functionApi.go#L224)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

Just nitpicking here, but this isn't the last pod executed, but the pod that was created last. Since we bind functions to pods late, the two can be different.  The actual code will work fine for now, I think, as a heuristic.
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 233 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r6-233:-Kxg3BEgktvN3HxcQYAZ:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/controller/functionApi.go#L233)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

`pods[0]` won't exist if there are no pods, and you'll get a runtime range check panic. So you have to handle the case where there are zero pods.
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 240 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r6-240:-Kxg3CckPOQMGCzVOdS5:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/controller/functionApi.go#L240)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

This is ok as an initial implementation, but we'll end up reading logs into memory. We should stream them to the response (using `io.Copy` would work, I think). Please fix it if it's not too complex, but otherwise file an issue to come back to later.
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 205 at r7](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r7-205:-Kxg2nXcp34iqeULmP1p:b-896fix) ([raw file](https://github.com/fission/fission/blob/9be1f29b27be795876067bd501d821712ddc5362/controller/functionApi.go#L205)):*
<details><summary><i>Previously, vishal-biyani (Vishal) wrote…</i></summary><blockquote>

@soamvasani I will create an issue so we can track this.
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 215 at r7](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r7-215:-Kxg35jJZhxwFIknpgy3:b-896fix) ([raw file](https://github.com/fission/fission/blob/9be1f29b27be795876067bd501d821712ddc5362/controller/functionApi.go#L215)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

It would be better to move this to the init of the api object -- so you don't have to login to kubernetes api on every logs request
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 221 at r7](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r7-221:-Kxg37Mn9j22T709xXSD:b-896fix) ([raw file](https://github.com/fission/fission/blob/9be1f29b27be795876067bd501d821712ddc5362/controller/functionApi.go#L221)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

nit: comment is outdated in relation to the code... should just say function pods now
</blockquote></details>

Done.

---

*[controller/functionApi.go, line 257 at r7](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9Y-r7-257:-Kxg5_A_Sz2ii8FaFkOq:b-896fix) ([raw file](https://github.com/fission/fission/blob/9be1f29b27be795876067bd501d821712ddc5362/controller/functionApi.go#L257)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

I think this should move to be before the `io.Copy` call. Otherwise we won't close podLogs if there's an error writing to `w`.
</blockquote></details>

Done.

---

*[fission/common.go, line 64 at r2](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9U-r2-64:-Kxg3F9HYg10xJ7sZ7tr:b-896fix) ([raw file](https://github.com/fission/fission/blob/ca052f4a8fea13d20e9b2a0e29de28a6075e0d95/fission/common.go#L64)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

Can you remove the comment too if it's not relevant?
</blockquote></details>

Done.

---

*[fission/common.go, line 73 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9U-r6-73:-Kxg52wqRJY6pNjIWHMX:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/common.go#L73)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

checkErr(err, "create HTTP request")

checkErr prints "Failed to <msg>" where msg is the second argument.
</blockquote></details>

Done.

---

*[fission/common.go, line 78 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9U-r6-78:-Kxg54JMmgqEzBW51RvX:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/common.go#L78)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

We should fail completely instead of just warning and skipping the header. The semantics of the request could be arbitrarily different when a header is skipped -- we don't know anything about the function.
</blockquote></details>

Done.

---

*[fission/common.go, line 87 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9U-r6-87:-Kxg55e8VrgPfDgaKky-:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/common.go#L87)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

Log the error `err`
</blockquote></details>

Done.

---

*[fission/function.go, line 479 at r2](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r2-479:-Kxg5UBnMoztWtALtdVF:b-896fix) ([raw file](https://github.com/fission/fission/blob/ca052f4a8fea13d20e9b2a0e29de28a6075e0d95/fission/function.go#L479)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

The retry is good, but the first request attempt can be sent without waiting (in other words the sleep should be after the request attempt)
</blockquote></details>

Done.

---

*[fission/function.go, line 482 at r2](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r2-482:-Kxg5SEVWJzRSZabaHz7:b-896fix) ([raw file](https://github.com/fission/fission/blob/ca052f4a8fea13d20e9b2a0e29de28a6075e0d95/fission/function.go#L482)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

Don't ignore errors
</blockquote></details>

Done.

---

*[fission/function.go, line 492 at r2](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r2-492:-Kxg5R8ViNkha2KTwYxd:b-896fix) ([raw file](https://github.com/fission/fission/blob/ca052f4a8fea13d20e9b2a0e29de28a6075e0d95/fission/function.go#L492)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

handle errors
</blockquote></details>

Done.

---

*[fission/function.go, line 493 at r2](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r2-493:-Kxg5QgqSfq4TU4Fp4Qq:b-896fix) ([raw file](https://github.com/fission/fission/blob/ca052f4a8fea13d20e9b2a0e29de28a6075e0d95/fission/function.go#L493)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

print status code too? Something like `Error calling function <function>: <code> <body>`
</blockquote></details>

Done.

---

*[fission/function.go, line 495 at r2](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r2-495:-Kxg5JByRtUG0ynCZNIk:b-896fix) ([raw file](https://github.com/fission/fission/blob/ca052f4a8fea13d20e9b2a0e29de28a6075e0d95/fission/function.go#L495)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

don't ignore errors
</blockquote></details>

Done.

---

*[fission/function.go, line 498 at r2](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r2-498:-Kxg5IJdIJl_sx_h1lnL:b-896fix) ([raw file](https://github.com/fission/fission/blob/ca052f4a8fea13d20e9b2a0e29de28a6075e0d95/fission/function.go#L498)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

suggest removing
</blockquote></details>

Done.

---

*[fission/function.go, line 141 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-141:-Kxg57IkPS6LYbsJUE4-:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L141)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

This function doesn't just get logs, it prints them too -- it should be called "show pod logs" or "print pod logs" or something like that
</blockquote></details>

Done.

---

*[fission/function.go, line 148 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-148:-Kxg58jCadMnb649kdNd:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L148)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

error message grammar, `checkErr(err, "parse server URL")`
</blockquote></details>

Done.

---

*[fission/function.go, line 152 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-152:-Kxg596RqPhQzGskMC1t:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L152)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

error message grammar
</blockquote></details>

Done.

---

*[fission/function.go, line 154 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-154:-Kxg59nyiigRW9BQtF7A:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L154)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

How did you choose the timeout? Is it what kubectl does? 
</blockquote></details>

Done.

---

*[fission/function.go, line 167 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-167:-Kxg5BQ-7hdHTn43Wv0g:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L167)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

I'm confused about this. What does the body look like? Why use json.Unmarshal for just a string? Can you simply do `base64.StdEncoding.Decode(body)`?
</blockquote></details>

Done.

---

*[fission/function.go, line 169 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-169:-Kxg5BsZiARVhpDJlZfU:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L169)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

when would this happen?
</blockquote></details>

Done.

---

*[fission/function.go, line 172 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-172:-Kxg5AxESGAZdL8kvPVb:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L172)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

Remove "Data="
</blockquote></details>

Done.

---

*[fission/function.go, line 549 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-549:-Kxg5TZvofnqL2ErCE4L:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L549)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

I don't think this command needs retries, let's keep it simple and the user can retry if they want.
</blockquote></details>

Done.

---

*[fission/function.go, line 569 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-569:-Kxg5PoKVLF6HO_OOJDI:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L569)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

See other uses of checkErr -- this doesn't fit the grammar of the error message. Also, this error is not about the function test failing, it's about failing to read the http response
</blockquote></details>

Done.

---

*[fission/function.go, line 572 at r6](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r6-572:-Kxg5OoAy9qhKwuYoK9J:b-896fix) ([raw file](https://github.com/fission/fission/blob/3ef8396e8ec84d954e10628eaaa3da673ff5cc91/fission/function.go#L572)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

Don't use `bool` for error checking in Go -- use an `error`.
</blockquote></details>

Done.

---

*[fission/function.go, line 159 at r7](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r7-159:-Kxg5ASfGS00mJx14ESm:b-896fix) ([raw file](https://github.com/fission/fission/blob/9be1f29b27be795876067bd501d821712ddc5362/fission/function.go#L159)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

s/POD/pod/
</blockquote></details>

Done.

---

*[fission/function.go, line 541 at r7](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r7-541:-Kxg5T5UQ_1qhhTms5Sb:b-896fix) ([raw file](https://github.com/fission/fission/blob/9be1f29b27be795876067bd501d821712ddc5362/fission/function.go#L541)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

this function could be written more simply without this flag, right?
</blockquote></details>

Done.

---

*[fission/function.go, line 544 at r7](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r7-544:-Kxg5Sg9GA9msZXnNutx:b-896fix) ([raw file](https://github.com/fission/fission/blob/9be1f29b27be795876067bd501d821712ddc5362/fission/function.go#L544)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

3xx are not really errors, right? I don't think we should call them errors. This should be 400...
</blockquote></details>

Done.

---

*[fission/function.go, line 549 at r7](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9V-r7-549:-Kxg5Rruh-d6Nlhe0QAd:b-896fix) ([raw file](https://github.com/fission/fission/blob/9be1f29b27be795876067bd501d821712ddc5362/fission/function.go#L549)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

just return here instead of setting a flag?
</blockquote></details>

Done.

---

*[fission/main.go, line 62 at r2](https://reviewable.io:443/reviews/fission/fission/355#-KxUTmgIRu2QafcV2t9W-r2-62:-Kxg5HlIUXakETNFMz8D:b2nhktp) ([raw file](https://github.com/fission/fission/blob/ca052f4a8fea13d20e9b2a0e29de28a6075e0d95/fission/main.go#L62)):*
<details><summary><i>Previously, soamvasani (Soam Vasani) wrote…</i></summary><blockquote>

I don't think we need both. Just test (or just invoke) is fine, with all the functionality in one command.
</blockquote></details>

OK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/355)*
<!-- Sent from Reviewable.io -->

soamvasani commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r6, 1 of 3 files at r7, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fission/fission/355#-:-Kxku7mKdC_VZ79nDXup:bnfp4nl)*
<!-- Sent from Reviewable.io -->
